### PR TITLE
Accept an HttpUrl in the endpoint builder method.

### PR DESCRIPTION
**Pull request by JakeWharton**
Also add tests for all the builder methods and their counterparts in the final Retrofit class.

### CODE REVIEW

**Comment by JakeWharton** [comment]
Refs #891.

File: retrofit/src/main/java/retrofit/Endpoint.java [hunk 1]
```diff
@@ -2,29 +2,13 @@
...
-      throw new IllegalArgumentException("Invalid URL: " + url);
-    }
-    return new Endpoint() {
-      @Override public HttpUrl url() {
-        return httpUrl;
-      }
-    };
-  }
-
+public interface Endpoint {
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/Endpoint.java]: Do we need an interface? We could just use the `HttpUrl`?

Comment (swankjesse) [inline retrofit/src/main/java/retrofit/Endpoint.java]: (though then it isn't mutable)

Comment (swankjesse) [inline retrofit/src/main/java/retrofit/Endpoint.java]: (and cannot be computed on-demand)

Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/Endpoint.java]: Yeah it needs to be able to change over time. We use it internally and there's a few external consumers of the behavior.

**Pull request merged**